src/wallet/scriptpubkeyman.cpp

```diff
@@ -1817,6 +1817,19 @@ bool DescriptorScriptPubKeyMan::TopUp(unsigned int size)
                 m_wallet_descriptor.cache.CacheDerivedExtPubKey(derived_xpub_map_pair.first, derived_xpub_pair.first, derived_xpub_pair.second);
             }
         }
+        for (const auto& lh_xpub_pair : temp_cache.GetCachedLastHardenedExtPubKeys()) {
+            CExtPubKey xpub;
+            if (m_wallet_descriptor.cache.GetCachedLastHardenedExtPubKey(lh_xpub_pair.first, xpub)) {
+                if (xpub != lh_xpub_pair.second) {
+                    throw std::runtime_error(std::string(__func__) + ": New cached last hardened xpub does not match already cached last hardened xpub");
+                }
+                continue;
+            }
+            if (!batch.WriteDescriptorLastHardenedCache(lh_xpub_pair.second, id, lh_xpub_pair.first)) {
```

Comment:
`LoadWallet`, which is called as part of wallet creation, will always call `CWallet::UpgradeDescriptorCache`. Since a newly created wallet is never encrypted at that stage, and has no `ScriptPubKeyMan`s, `UpgradeDescriptorCache` will just set the flag.